Supplier Scorecard

bcornwellmott opened the pull request:
This PR is obviously a lot more than 30 lines. Most of the lines are from creating the doctypes, but there's also a lot of code. This feature is hidden from users for now, as there's still a lot of functionality to implement.

The main things still to do are:

- [x] Debug any errors
- [x] Write tests
- [x] Create the default records (variables, criteria, etc)
- [x] Test custom python variables
- [x] Tag all doctypes as "beta"
- [x] Fix Codacy erros

bcornwellmott committed:
Initial start of scorecard docs

bcornwellmott committed:
Got basic functionality working

bcornwellmott committed:
Fixed doc names and added key functions

bcornwellmott committed:
Basic functional version minus Actions

bcornwellmott committed:
Hide scorecard docs until functional

rmehta commented:
@bcornwellmott 👍  Can you share a design, so that we can quickly review anything that might go against guidelines. Thanks.

rmehta committed:
Merge branch 'develop' into scorecard

rmehta commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,177 @@
+		self.validate_criteria_weights()
+		self.calculate_total_score()
+		self.update_standing()
+		
```
why so much whitespace?

rmehta commented:
@bcornwellmott please add basic screenshots and explanations of what is included in the PR, esp the Forms.

Ok with not going ahead with the 30 lines rule for this one.

Please reopen with screenshots and docs.

bcornwellmott commented:
Full screenshots and explanations are here:
https://discuss.erpnext.com/t/new-feature-supplier-scorecard/24520

bcornwellmott committed:
Created supplier scorecard documentation

bcornwellmott committed:
Added default variables and standings. Added restrictions for PO + RFQ

bcornwellmott committed:
Automatic daily scorecard creation + on save

bcornwellmott committed:
Added warning for PO nd RFQs

bcornwellmott commented:
I've added some user manual documentation, added warn for RFQs and POs (based on forum feedback), implemented all the action checkboxes (for warn and prevent), and automated the scorecard period creation (daily + on save).

I still need to write the test cases, and I've hidden the notify fields until I've added the functionality to actual notify the supplier / employee.

bcornwellmott committed:
Minor fixes for codepy, automatically add variables for criteria, fix hooks.py typo

bcornwellmott committed:
Added tests, fixed codacy formatting, small improvements

bcornwellmott committed:
Merge branch 'develop' into scorecard

bcornwellmott committed:
Fixed test bug w/ criteria. Codacy cleanup

bcornwellmott committed:
Merge branch 'scorecard' of https://github.com/bcornwellmott/erpnext into scorecard

bcornwellmott committed:
Fixed codacy issues. Fixed sticky criteria

bcornwellmott commented:
Added tests and tried to clean up the code. I'm not sure all of the codacy suggestions fit with how the rest of ERPNext is coded. Specifically:
 - camel_case
 - shorthand
It seems that the rest of ERPNext is formatted this way, so I think these checks should be disabled because they contradict all the code in ERPNext.

netchampfaris commented:
I don't know why but codacy is picking up @nabinhait 's codacy settings, these errors don't show up in the official **frappe** codacy account

nabinhait commented:
@netchampfaris I have removed **frappe/erpnext** from my codacy account and setup **nabinhait/erpnext** account.

bcornwellmott committed:
Fixed bug with period search. Remove blank variable child.

bcornwellmott commented:
Here is a gif of how to create:
![scorecard](https://user-images.githubusercontent.com/16806041/28030212-a5da6c40-6557-11e7-8397-8dff107c00c7.gif)

rmehta commented:
@bcornwellmott might be a good idea to pre-set the scorecard so that the user can just adjust the criteria.

Also for how many periods will the cards be generated? The long message UX can be improved!

bcornwellmott commented:
The scorecards get generated from when the supplier was first created to the current date (then gets generated for every subsequent period through a scheduled task). I'll clean up the message UX. 

I wanted to leave the criteria alone (not create any by default) because it's something that needs to be tailored to your organization. What I can do is automatically load all of them in, so you save some clicks when setting it up (I'll do this for standings as well). I can also add some more details in the documentation to explain how to generate the formulas, with some possible examples (starting points).

bcornwellmott committed:
Updated docs, automatically add criteria and standings, clean up period create message

rmehta commented:
cool. I am okay going ahead, can you set "beta": 1 in all the doctypes so that users know that this functionality is new. We can remove beta after 4-6 weeks once there is user feedback.

bcornwellmott committed:
Uncommented test, set docs to beta

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.json:
```diff
@@ -0,0 +1,731 @@
+   "label": "Evaluation Period", 
+   "length": 0, 
+   "no_copy": 0, 
+   "options": "Per Week\nPer Month\nPer Year", 
```
Use `Monthly/Weekly/Yearly` to make it consistent with other documents.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+		if (frm.doc.indicator_color !== "")
+		{
+			frm.set_indicator_formatter("status", function(doc) {
+				return doc.indicator_color.toLowerCase();
```
Found that `indicator_color ` is hidden. If it is not used delete the field and code.

bcornwellmott replied:
Indicator color is used for the list view. Pulled from the standing.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+
+	onload: function(frm) {
+		if (frm.doc.indicator_color !== "")
+		{
```
style guide: `if (frm.doc.indicator_color !== "") {`

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+			});
+		}
+		if (frm.doc.__unsaved == 1)
+		{
```
style guide: We use this brace at the end of `if` condition in the same line. Please fix this everywhere.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+		}
+		if (frm.doc.__unsaved == 1)
+		{
+			frappe.call({
```
Put this in a separate function

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+					refresh_field("criteria");
+				}
+			});
+			frappe.call({
```
Put this in a separate function

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.js:
```diff
@@ -0,0 +1,161 @@
+			frm.dashboard.heatmap.setLegend([0,20,40,60,80,101],["#991600","#169900"]);
+		}
+	},
+	generate_scorecards:function(frm) {
```
Found this button hidden, if not used, delete both code and field

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.json:
```diff
@@ -0,0 +1,731 @@
+   "in_global_search": 0, 
+   "in_list_view": 0, 
+   "in_standard_filter": 0, 
+   "label": "Warn POs", 
```
Use full name of doctype and more descriptive label. I don't understand on which condition it will warn.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.json:
```diff
@@ -0,0 +1,731 @@
+   "in_global_search": 0, 
+   "in_list_view": 0, 
+   "in_standard_filter": 0, 
+   "label": "Employee Link", 
```
Use employee instead of employee_name

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+			for c2 in self.standings:
+				if c1 != c2:
+					if (c1.max_grade > c2.min_grade and c1.min_grade < c2.max_grade):
+						throw(_('Overlap in scoring between ' + c1.standing_name + ' and ' + c2.standing_name))
```
Use `_('Overlap in scoring between {0} and {1}).format(c1.standing_name, c2.standing_name)` to make it translation friendly.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+
+		# Check that there are no missing scores
+		score = 0
+		for c1 in self.standings:
```
Why a separate loop, it can be written in the above loop.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+				if c2.min_grade == score:
+					score = c2.max_grade
+		if score < 100 :
+			throw(_('Unable to find score starting at ' + str(score) + '. You need to have standing scores covering 0 to 100'))
```
use `.format()`

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+				frappe.db.set_value("Supplier", self.supplier, "prevent_pos", self.prevent_pos)
+				frappe.db.set_value("Supplier", self.supplier, "prevent_rfqs", self.prevent_rfqs)
+				frappe.db.set_value("Supplier", self.supplier, "warn_rfqs", self.warn_rfqs)
+				frappe.db.set_value("Supplier", self.supplier, "warn_pos", self.warn_pos)
```
as the fieldname is same in source and target, you can write like this:
```
for fieldname in ('prevent_pos', 'prevent_rfqs', 'warn_rfqs'):
  self.set(fieldname, standing.get(fieldname))
  frappe.db.set_value("Supplier", self.supplier, fieldname, self.get(fieldname))
```

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+				frappe.db.set_value("Supplier", self.supplier, "warn_pos", self.warn_pos)
+
+@frappe.whitelist()
+def timeline_data(doctype, name):
```
`get_timeline_data()`?

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+			{"scs": scs.name}, as_dict=1)
+
+	for sc in scorecards:
+		sc = frappe.get_doc('Supplier Scorecard Period', sc.name)
```
To fetch 2/3 fields value, try to avoid `get_doc`. Use `get_value` instead.

`start_date, end_date = frappe.db.get_value('Supplier Scorecard Period', sc.name, ['start_date', 'end_date'])`

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+
+def make_install_records():
+	install_docs = [
+		{"param_name":"total_accepted_items","name":"Total Accepted Items","doctype":"Supplier Scorecard Variable","variable_label":"Total Accepted Items","path":"get_total_accepted_items"},
```
Please use space after comma(, ) and colon(: ).

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+		{"param_name":"total_accepted_amount","name":"Total Accepted Amount","doctype":"Supplier Scorecard Variable","variable_label":"Total Accepted Amount","path":"get_total_accepted_amount"},
+		{"param_name":"total_rejected_items","name":"Total Rejected Items","doctype":"Supplier Scorecard Variable","variable_label":"Total Rejected Items","path":"get_total_rejected_items"},
+		{"param_name":"total_rejected_amount","name":"Total Rejected Amount","doctype":"Supplier Scorecard Variable","variable_label":"Total Rejected Amount","path":"get_total_rejected_amount"},
+		{"param_name":"total_received_items","name":"Total Received Items","doctype":"Supplier Scorecard Variable","variable_label":"Total Received Items","path":"get_total_received_items"},
```
As naming is based on `variable_label`, you don't need to pass `name` explicitely.

Also, you can use doctype in common, to cleanup the data, it's really unreadable. Also break the lines after 80 chars.

nabinhait commented on erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py:
```diff
@@ -0,0 +1,237 @@
+		end_date = add_days(add_years(start_date,1), -1)
+	return end_date
+
+def make_install_records():
```
the function name can be `make_default_records`

nabinhait commented:
Addition to inline comments, also there is an error while creating a new Criteria from Supplier Scorecard record.

<img width="857" alt="screen shot 2017-07-17 at 1 33 07 pm" src="https://user-images.githubusercontent.com/836784/28261100-8061b9b0-6afb-11e7-90ef-eff0ff3c493f.png">


Closing this PR, please reopen after fixing.

bcornwellmott committed:
Fix for nabinhait review

bcornwellmott commented:
Implemented the changes you suggested @nabinhait . Please let me know if there's anything else.

bcornwellmott committed:
Merge branch 'develop' into scorecard

bcornwellmott committed:
Fix codacy issue. Fix dict assignment for records

bcornwellmott committed:
Merge branch 'scorecard' of https://github.com/bcornwellmott/erpnext into scorecard

bcornwellmott committed:
Merge branch 'develop' into scorecard

nabinhait committed:
Merge branch 'develop' into scorecard
